File: java-examples/src/main/java/rpccalls/EditOffer.java
package rpccalls;

import bisq.proto.grpc.EditOfferRequest;
import bisq.proto.grpc.OffersGrpc;
import io.grpc.ManagedChannelBuilder;

import static bisq.proto.grpc.EditOfferRequest.EditType.*;
import static bisq.proto.grpc.EditOfferRequest.newBuilder;
import static java.lang.System.out;

public class EditOffer extends BaseJavaExample {

    public static void main(String[] args) {
        try {
            var channel = ManagedChannelBuilder.forAddress("localhost", 9998).usePlaintext().build();
            addChannelShutdownHook(channel);
            var credentials = buildCallCredentials(getApiPassword());
            var stub = OffersGrpc.newBlockingStub(channel).withCallCredentials(credentials);

            var disableOfferRequest = createDisableOfferRequest();
            stub.editOffer(disableOfferRequest);
            out.println("Offer is disabled (removed from peers' offer book).");
            waitForOfferToBeRepublished(2_500);

            var editFixedPriceRequest = createEditFixedPriceRequest();
            stub.editOffer(editFixedPriceRequest);
            out.println("Offer has been re-published with new fixed-price.");
            waitForOfferToBeRepublished(2_500);

            var editFixedPriceAndEnableRequest = createEditFixedPriceAndEnableRequest();
            stub.editOffer(editFixedPriceAndEnableRequest);
            out.println("Offer has been re-published with new fixed-price, and enabled.");
            waitForOfferToBeRepublished(2_500);

            var editPriceMarginRequest = createEditPriceMarginRequest();
            stub.editOffer(editPriceMarginRequest);
            out.println("Offer has been re-published with new price margin.");
            waitForOfferToBeRepublished(2_500);

            var editTriggerPriceRequest = createEditTriggerPriceRequest();
            stub.editOffer(editTriggerPriceRequest);
            out.println("Offer has been re-published with new trigger price.");
        } catch (Throwable t) {
            handleError(t);
        }
    }

    private static EditOfferRequest createDisableOfferRequest() {
        return newBuilder()
                .setId("83e8b2e2-51b6-4f39-a748-3ebd29c22aea")
                .setEditType(ACTIVATION_STATE_ONLY)
                .setEnable(0)   // -1 = ignore this parameter, 0 = disable offer, 1 = enable offer
                .build();
    }

    private static EditOfferRequest createEditFixedPriceRequest() {
        return newBuilder()
                .setId("83e8b2e2-51b6-4f39-a748-3ebd29c22aea")
                .setEditType(FIXED_PRICE_ONLY)
                .setPrice("30000.99")
                .build();
    }

    private static EditOfferRequest createEditFixedPriceAndEnableRequest() {
        return newBuilder()
                .setId("83e8b2e2-51b6-4f39-a748-3ebd29c22aea")
                .setEditType(FIXED_PRICE_AND_ACTIVATION_STATE)
                .setPrice("30000.99")
                .setEnable(1)
                .build();
    }

    private static EditOfferRequest createEditPriceMarginRequest() {
        return newBuilder()
                .setId("83e8b2e2-51b6-4f39-a748-3ebd29c22aea")
                .setEditType(MKT_PRICE_MARGIN_ONLY)
                .setUseMarketBasedPrice(true)
                .setMarketPriceMarginPct(2.00)  // 2.00%
                .build();
    }

    private static EditOfferRequest createEditTriggerPriceRequest() {
        return newBuilder()
                .setId("83e8b2e2-51b6-4f39-a748-3ebd29c22aea")
                .setEditType(TRIGGER_PRICE_ONLY)
                .setTriggerPrice("29000.00")    // Trigger price is disabled when set to "0".
                .build();
    }

    private static void waitForOfferToBeRepublished(long ms) {
        try {
            // Wait for edited offer to be removed from offer-book, edited, and re-published.
            Thread.sleep(ms);
        } catch (InterruptedException ex) {
            // ignored
        }
    }
}

File: python-examples/rpccalls/edit_offer.py
# from getpass import getpass
import time
from builtins import print

import grpc

import bisq.api.grpc_pb2 as bisq_messages
import bisq.api.grpc_pb2_grpc as bisq_service

EDITED_USD_OFFER_ID = '44736-16df6819-d98b-4f13-87dd-50087c464fac-184'


def main():
    grpc_channel = grpc.insecure_channel('localhost:9998')
    grpc_service_stub = bisq_service.OffersStub(grpc_channel)
    api_password: str = 'xyz'  # getpass("Enter API password: ")
    try:
        edit_offer_request = disable_offer_request()
        edit_offer_response = grpc_service_stub.EditOffer.with_call(edit_offer_request,
                                                                    metadata=[('password', api_password)])
        print('Offer is disabled.  Rpc response: ' + str(edit_offer_response))
        time.sleep(4)  # Wait for new offer preparation and wallet updates before creating another offer.

        edit_offer_request = enable_offer_request()
        edit_offer_response = grpc_service_stub.EditOffer.with_call(edit_offer_request,
                                                                    metadata=[('password', api_password)])
        print('Offer is enabled.  Rpc response: ' + str(edit_offer_response))
        time.sleep(4)  # Wait for new offer preparation and wallet updates before creating another offer.

        edit_offer_request = edit_fixed_price_request()
        edit_offer_response = grpc_service_stub.EditOffer.with_call(edit_offer_request,
                                                                    metadata=[('password', api_password)])
        print('Offer fixed-price has been changed.  Rpc response: ' + str(edit_offer_response))
        time.sleep(4)  # Wait for new offer preparation and wallet updates before creating another offer.

        edit_offer_request = edit_fixed_price_and_enable_request()
        edit_offer_response = grpc_service_stub.EditOffer.with_call(edit_offer_request,
                                                                    metadata=[('password', api_password)])
        print('Offer fixed-price has been changed, and offer is enabled.  Rpc response: ' + str(edit_offer_response))
        time.sleep(4)  # Wait for new offer preparation and wallet updates before creating another offer.

        # Change the fixed-price offer to a mkt price margin based offer
        edit_offer_request = edit_price_margin_request()
        edit_offer_response = grpc_service_stub.EditOffer.with_call(edit_offer_request,
                                                                    metadata=[('password', api_password)])
        print('Fixed-price offer is not a mkt price margin based offer.  Rpc response: ' + str(edit_offer_response))
        time.sleep(4)  # Wait for new offer preparation and wallet updates before creating another offer.

        # Set the trigger-price on a mkt price margin based offer
        edit_offer_request = edit_trigger_price_request()
        edit_offer_response = grpc_service_stub.EditOffer.with_call(edit_offer_request,
                                                                    metadata=[('password', api_password)])
        print('Offer trigger price is set.  Rpc response: ' + str(edit_offer_response))

    except grpc.RpcError as rpc_error:
        print('gRPC API Exception: %s', rpc_error)


def disable_offer_request():
    return bisq_messages.EditOfferRequest(
        id=EDITED_USD_OFFER_ID,
        edit_type=bisq_messages.EditOfferRequest.EditType.ACTIVATION_STATE_ONLY,
        enable=0)  # If enable=-1: ignore enable param, enable=0: disable offer, enable=1: enable offer


def enable_offer_request():
    return bisq_messages.EditOfferRequest(
        id=EDITED_USD_OFFER_ID,
        edit_type=bisq_messages.EditOfferRequest.EditType.ACTIVATION_STATE_ONLY,
        enable=1)  # If enable=-1: ignore enable param, enable=0: disable offer, enable=1: enable offer


def edit_fixed_price_request():
    return bisq_messages.EditOfferRequest(
        id=EDITED_USD_OFFER_ID,
        price='42000.50',
        edit_type=bisq_messages.EditOfferRequest.EditType.FIXED_PRICE_ONLY,
        enable=-1)  # If enable=-1: ignore enable param, enable=0: disable offer, enable=1: enable offer


def edit_fixed_price_and_enable_request():
    return bisq_messages.EditOfferRequest(
        id=EDITED_USD_OFFER_ID,
        price='43000.50',
        edit_type=bisq_messages.EditOfferRequest.EditType.FIXED_PRICE_AND_ACTIVATION_STATE,
        enable=1)  # If enable=-1: ignore enable param, enable=0: disable offer, enable=1: enable offer


def edit_price_margin_request():
    return bisq_messages.EditOfferRequest(
        id=EDITED_USD_OFFER_ID,
        use_market_based_price=True,
        market_price_margin_pct=5.00,
        edit_type=bisq_messages.EditOfferRequest.EditType.MKT_PRICE_MARGIN_ONLY,
        enable=-1)  # If enable=-1: ignore enable param, enable=0: disable offer, enable=1: enable offer


def edit_trigger_price_request():
    return bisq_messages.EditOfferRequest(
        id=EDITED_USD_OFFER_ID,
        trigger_price='40000.0000',
        edit_type=bisq_messages.EditOfferRequest.EditType.TRIGGER_PRICE_ONLY,
        enable=-1)  # If enable=-1: ignore enable param, enable=0: disable offer, enable=1: enable offer


if __name__ == '__main__':
    main()
